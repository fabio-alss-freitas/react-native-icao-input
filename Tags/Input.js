import React from "react"
import { View, TextInput } from "react-native"

import styles from "./styles"

const Input = (props) => {
	const {
		value,
		onChangeText,
		onSubmitEditing,
		onKeyPress,
		inputStyle,
		inputContainerStyle,
		textInputProps,
	} = props

	return (
		<View style={[styles.textInputContainer, inputContainerStyle]}>
			<TextInput
				{...textInputProps}
				style={[styles.textInput, inputStyle]}
				value={value}
				onChangeText={onChangeText}
				onSubmitEditing={onSubmitEditing}
				onKeyPress={onKeyPress}
				underlineColorAndroid="transparent"
			/>
		</View>
	)
}

export { Input }
export default Input
