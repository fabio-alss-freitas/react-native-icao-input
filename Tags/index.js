import React, { useState, useImperativeHandle, useEffect } from "react"
import PropTypes from "prop-types"
import { View } from "react-native"

import Tag from "./Tag"
import Input from "./Input"
import styles from "./styles"

const Tags = React.forwardRef((props, ref) => {
	useImperativeHandle(
		ref,
		() => ({
			clear: () => clearTags(),
		}),
		[ref]
	)

	const {
		containerStyle,
		style,
		readonly,
		maxNumberOfTags,
		tagContainerStyle,
		tagTextStyle,
		deleteTagOnPress,
		onTagPress = () => {},
		renderTag,
		initialTags = [],
		initialText,
		onChangeTags = () => {},
	} = props
	const [tags, setTags] = useState(initialTags)
	const [text, setText] = useState(initialText)

	useEffect(() => {
		onChangeTags(tags || [])
	}, [tags])

	const showLastTag = () => {
		const aux = [...tags]
		setTags(aux?.slice(0, -1) || [])
		setText(aux?.slice(-1)[0]?.slice(0, -1) || "")
	}

	const addTag = (text) => {
		setTags((prev) => [...prev, text.trim()])
		setText("")
	}

	const clearTags = () => {
		setTags([])
		setText("")
	}

	const onChangeText = (text) => {
		if (text.length === 0) {
			showLastTag()
		} else if (text.length == 4 && !(tags.indexOf(text.slice(0, -1).trim()) > -1)) {
			addTag(text.replace(/[^a-zA-Z0-9]/g, ""))
		} else {
			setText(text.replace(/[^a-zA-Z0-9]/g, ""))
		}
	}

	const onKeyPress = ({ nativeEvent }) => {
		if (nativeEvent.key === "Backspace" && text?.length === 0) {
			showLastTag()
		}
	}

	return (
		<View style={[styles.container, containerStyle, style]}>
			{tags.map((tag, index) => {
				const tagProps = {
					tag,
					index,
					deleteTagOnPress,
					onPress: (event) => {
						event?.persist()
						if (deleteTagOnPress && !readonly) {
							setTags((prev) => [...prev.slice(0, index), ...prev.slice(index + 1)])
							onTagPress(index, tag, event, true)
						} else {
							onTagPress(index, tag, event, false)
						}
					},
					tagContainerStyle,
					tagTextStyle,
				}

				return renderTag(tagProps)
			})}

			{!readonly && maxNumberOfTags > tags.length && (
				<Input value={text} onChangeText={onChangeText} onKeyPress={onKeyPress} {...props} />
			)}
		</View>
	)
})

Tags.defaultProps = {
	initialTags: [],
	initialText: " ",
	createTagOnString: [",", " "],
	createTagOnReturn: false,
	readonly: false,
	deleteTagOnPress: true,
	maxNumberOfTags: Number.POSITIVE_INFINITY,
	renderTag: ({ tag, index, ...rest }) => <Tag key={`${tag}-${index}`} label={tag} {...rest} />,
}

Tags.propTypes = {
	initialText: PropTypes.string,
	initialTags: PropTypes.arrayOf(PropTypes.string),
	createTagOnString: PropTypes.array,
	createTagOnReturn: PropTypes.bool,
	onChangeTags: PropTypes.func,
	readonly: PropTypes.bool,
	maxNumberOfTags: PropTypes.number,
	deleteTagOnPress: PropTypes.bool,
	renderTag: PropTypes.func,
	/* style props */
	containerStyle: PropTypes.any,
	style: PropTypes.any,
	inputContainerStyle: PropTypes.any,
	inputStyle: PropTypes.any,
	tagContainerStyle: PropTypes.any,
	tagTextStyle: PropTypes.any,
	textInputProps: PropTypes.object,
}

export { Tags }
export default Tags
