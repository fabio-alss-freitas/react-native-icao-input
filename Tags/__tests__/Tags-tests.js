import React from "react";
import renderer from "react-test-renderer";
import { Text } from "react-native";

import Tags from "../../";

describe("Tags", () => {
  const noop = function() {};

  it("should render props correctly", () => {
    const tree = renderer
      .create(
        <Tags
          initialText="monkey gland sauce"
          initialTags={["tomato sauce", "mustard", "mayo"]}
          onChangeTags={noop}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("renderTag prop", () => {
    it("allows me to render a custom tag with a function", () => {
      const renderTag = jest.fn(({ tag }) => <Text key={tag}>{tag}</Text>);

      const tree = renderer
        .create(
          <Tags
            initialText=""
            initialTags={["palm", "oil", "sucks"]}
            onChangeTags={noop}
            renderTag={renderTag}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("textInputProps", () => {
    const tree = renderer
      .create(
        <Tags
          initialText="monkey gland sauce"
          textInputProps={{ placeholder: "testy text" }}
          initialTags={["tomato sauce", "mustard", "mayo"]}
          onChangeTags={noop}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
