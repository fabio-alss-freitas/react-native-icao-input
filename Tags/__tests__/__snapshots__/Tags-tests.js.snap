// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      tomato sauce
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      mustard
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      mayo
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 16,
          "flex": 1,
          "height": 32,
          "margin": 4,
          "minWidth": 100,
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="testy text"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "flex": 1,
            "fontSize": 13,
            "height": 32,
            "margin": 0,
            "padding": 0,
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value="monkey gland sauce"
    />
  </View>
</View>
`;

exports[`Tags renderTag prop allows me to render a custom tag with a function 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
      },
      undefined,
      undefined,
    ]
  }
>
  <Text>
    palm
  </Text>
  <Text>
    oil
  </Text>
  <Text>
    sucks
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 16,
          "flex": 1,
          "height": 32,
          "margin": 4,
          "minWidth": 100,
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "flex": 1,
            "fontSize": 13,
            "height": 32,
            "margin": 0,
            "padding": 0,
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`Tags should render props correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      tomato sauce
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      mustard
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 16,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 12,
        "paddingRight": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13,
          },
          undefined,
        ]
      }
    >
      mayo
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 16,
          "flex": 1,
          "height": 32,
          "margin": 4,
          "minWidth": 100,
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "flex": 1,
            "fontSize": 13,
            "height": 32,
            "margin": 0,
            "padding": 0,
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value="monkey gland sauce"
    />
  </View>
</View>
`;
